src/rgw/rgw_common.cc

```diff
@@ -1467,3 +1478,78 @@ int rgw_parse_op_type_list(const string& str, uint32_t *perm)
   return parse_list_of_flags(op_type_mapping, str, perm);
 }
 
+static int match_internal(const char* pattern, const char* input)
+{
+  while(true) {
+    if (*pattern == '\0' && *input == '\0')
+      return 1;
+    if (*pattern == '\0' || *input == '\0')
+      return 0;
+    if (*pattern == '*' && *(pattern + 1) == '\0' && *input != '\0')
+      return 1;
+    if (*pattern == '*' && *(pattern + 1) == '\0' && *input == '\0')
+      return 0;
+    if (*pattern == *input || *pattern == '?') {
+      pattern++;
+      input++;
+      continue;
+    }
+    if (*pattern == '*') {
+      if (*(pattern + 1) == *input)
+        pattern++;
+      else
+        input++;
+      continue;
+    }
+    return 0;
+  }
+  return 0;
+}
+
+static void tokenize(string& text, vector<string>& result)
+{
+  char_separator<char> sep(":");
+  tokenizer<char_separator<char> > tokens(text, sep);
+  for ( auto it : tokens)
+  {
+    result.push_back(it);
+  }
+}
+
+int match(string& pattern, string& input, int flag)
+{
```

Comment:
We should not mutate our arguments. (That is, they should either be const std::string& or a boost::string_ref. I'd prefer the latter since we can pass in C-style strings that way without having to copy them.)